server/src/main/java/org/opensearch/cluster/routing/WeightedRoundRobin.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class WeightedRoundRobin<T> implements Iterator<T>, Iterable<T> {
+
+    private List<WeightedRoundRobin.Entity<T>> entities;
+    private int turn;
+    private int lastSelectedEntity;
+    private double currentWeight = 0;
+
+    public WeightedRoundRobin(List<WeightedRoundRobin.Entity<T>> entities) {
+        this.entities = entities;
+        this.turn = 0;
+        this.lastSelectedEntity = -1;
```

Comment:
can you pls help me understand how `lastSelectedEntity` is used ? As per my  understanding, we create this on every search request and don't reuse it . Is it correct ?

Reply:
Yeah, in our case we are not returning a single entity but all the entities in WRR manner. Thats why ```lastSelectedEntity``` is not considered and is always set to -1. If we see [this](http://kb.linuxvirtualserver.org/wiki/Weighted_Round-Robin_Scheduling) implementation where its returning a single entity, ```lastSelectedEntity``` can be passed in the parameter